### PR TITLE
WIP: Presenced login refine

### PR DESCRIPTION
**Pull request by alxvasilev**
Refined reconnect/login logic of presenced client. To test: connect with internet, then disable internet and wait for retries to start, then connect internet and confirm reconnect is done properly.

### CODE REVIEW

**Comment by sergiohs84** [comment]
trigger compilation

File: src/presenced.cpp [hunk 1]
```diff
@@ -281,10 +291,18 @@ Client::reconnect(const std::string& url)
...
-        return retry("presenced", [this](int no)
+        auto wptr = weakHandle();
+        if (mRetryCtrl)
         {
+            mRetryCtrl->abort();
+        }
+        mRetryCtrl.reset(createRetryController("presenced", [wptr, this](int no) -> Promise<void>
+        {
+            if (wptr.deleted())
+                return promise::Error("destroyed");
```
Comment (sergiohs84) [inline src/presenced.cpp]: Shouldn't we abort the retry controller in this case??

File: src/presenced.cpp [hunk 2]
```diff
@@ -342,21 +356,32 @@ Client::reconnect(const std::string& url)
...
+                auto errtype = (err.type() == ERRTYPE_MEGASDK) ? WS_ERRTYPE_DNS : err.type();
+                if (!mConnectPromise.done())
+                    mConnectPromise.reject(err.msg(), err.code(), errtype);
+                if (!mLoginPromise.done())
+                    mLoginPromise.reject(err.msg(), err.code(), errtype);
             });
             
             return mConnectPromise
-            .then([this]()
+            .then([this]() -> Promise<void>
```
Comment (sergiohs84) [inline src/presenced.cpp]: Don't we want here the `weakptr` protection??

File: src/presenced.cpp [hunk 3]
```diff
@@ -342,21 +356,32 @@ Client::reconnect(const std::string& url)
...
             {
                 mHeartbeatEnabled = true;
                 return login();
+            })
+            .fail([this](const promise::Error& err)
```
Comment (sergiohs84) [inline src/presenced.cpp]: And also here, the `weakptr`

File: src/presenced.cpp (hunk 1 above)
Comment (alxvasilev) [inline src/presenced.cpp]: That's what we do - `mRetryCtrl->abort()`, or did I misunderstand what you mean?

File: src/presenced.cpp (hunk 2 above)
Comment (alxvasilev) [inline src/presenced.cpp]: True

File: src/presenced.cpp (hunk 3 above)
Comment (alxvasilev) [inline src/presenced.cpp]: Also true

File: src/presenced.cpp (hunk 1 above)
Comment (sergiohs84) [inline src/presenced.cpp]: In line 302, where the `presence::Client` has been deleted, the code returns a promise. However, I've seen how heartbeats / retry controller results in new connection attempts even when the Client is already defunct.
With my comment, I'm just calling your attention to that point.

File: src/presenced.cpp [hunk 4]
```diff
@@ -191,8 +193,16 @@ void Client::onSocketClose(int errcode, int errtype, const std::string& reason)
 
     if (mConnState < kLoggedIn) //tell retry controller that the connect attempt failed
     {
-        assert(!mLoginPromise.done());
-        mConnectPromise.reject(reason, errcode, errtype);
+        assert(!mConnectPromise.succeeded());
```
Comment (sergiohs84) [inline src/presenced.cpp]: No answers for this one? :(

Comment (alxvasilev) [inline src/presenced.cpp]: I needed to think a bit :) You are right, I removed the assert.

File: src/presenced.cpp (hunk 1 above)
Comment (sergiohs84) [inline src/presenced.cpp]: Good news indeed! Thanks for the explanation

File: src/presenced.cpp [hunk 5]
```diff
@@ -342,21 +355,36 @@ Client::reconnect(const std::string& url)
...
             {
+                if (wptr.deleted())
+                    return promise::_Void();
                 mHeartbeatEnabled = true;
                 return login();
+            })
+            .fail([wptr, this](const promise::Error& err)
+            {
+                if (wptr.deleted())
+                    return;
```
Comment (sergiohs84) [inline src/presenced.cpp]: Don't we need to return a `promise::Error()` here?

File: src/presenced.cpp [hunk 6]
```diff
@@ -342,21 +355,36 @@ Client::reconnect(const std::string& url)
...
             })
-            .fail([this](const promise::Error& err)
+            .fail([wptr, this](const promise::Error& err)
             {
-                int errType =  (err.type() == ERRTYPE_MEGASDK) ? WS_ERRTYPE_DNS : err.type();
-                mConnectPromise.reject(err.msg(), err.code(), errType);
-                mLoginPromise.reject(err.msg(), err.code(), errType);
+                PRESENCED_LOG_ERROR("Connect error: %s", err.what());
+                if (wptr.deleted())
+                    return;
```
Comment (sergiohs84) [inline src/presenced.cpp]: Don't we need to return a promise::Error() here?

**Comment by sergiohs84** [comment]
I got this 2 errors:
a) `WARNING: Unhandled promise fail. Error: 'Websocket error -1 on operation "connect"', type: 1, code: 1`
b) `WARNING: Unhandled promise fail. Error: 'aborted', type: 1, code: 2`

File: src/presenced.cpp (hunk 6 above)
Comment (alxvasilev) [inline src/presenced.cpp]: Nop, in both places. You need to handle the error eventually, i.e. not return it further. If you don't intend to attach any further error handlers (like in this case), then you should not propagate the error further, or you will get the "unhanded promise fail" warning.

File: src/presenced.cpp (hunk 5 above)
Comment (alxvasilev) [inline src/presenced.cpp]: Same as above

File: src/presenced.cpp [hunk 7]
```diff
@@ -342,21 +355,36 @@ Client::reconnect(const std::string& url)
...
+                    return promise::_Void();
                 mHeartbeatEnabled = true;
                 return login();
+            })
+            .fail([wptr, this](const promise::Error& err)
+            {
+                if (wptr.deleted())
+                    return;
+                PRESENCED_LOG_ERROR("Error connecting or logging in: %s", err.what());
             });
```
Comment (alxvasilev) [inline src/presenced.cpp]: Weird, what's the warning? You use gcc, right?

File: src/presenced.cpp [hunk 8]
```diff
@@ -342,21 +355,36 @@ Client::reconnect(const std::string& url)
...
-                mLoginPromise.reject(err.msg(), err.code(), errType);
+                PRESENCED_LOG_ERROR("Connect error: %s", err.what());
+                if (wptr.deleted())
+                    return;
+                auto errtype = (err.type() == ERRTYPE_MEGASDK) ? WS_ERRTYPE_DNS : err.type();
+                if (!mConnectPromise.done())
+                    mConnectPromise.reject(err.msg(), err.code(), errtype);
+                if (!mLoginPromise.done())
+                    mLoginPromise.reject(err.msg(), err.code(), errtype);
             });
```
Comment (alxvasilev) [inline src/presenced.cpp]: Weird, what's the warning? You use gcc, right?

Comment (sergiohs84) [inline src/presenced.cpp]: Yes, gcc. The warning says something like _non-void function should return a value_

**Comment by sergiohs84** [comment]
This branch is totally outdated and won't be merged. Closing...

**Pull request closed without merging**